### PR TITLE
Try to unify block event

### PR DESCRIPTION
**Pull request by ci010**
This will decide the block event api. Please review carefully.

Related issue #33

### CODE REVIEW

**Review by mouse0w0 — COMMENTED** [review]
Should we rename click and activate behaviors? Click behaviors maybe meaning left, right or middle click.

**Comment by mouse0w0** [comment]
Any block event should be divided into 'Pre' and 'Post'. 'BlockXxxEvent.Pre' is cancellable and post before notifying block behavior, 'BlockXxxEvent.Post' is noncancellable and post after notifying block behavior.

**Comment by ci010** [comment]
If you consider left right middle, that's saying you stick game logic with specific keybinding, which is a different term. 

You have to consider console controller button then.

Under considerations above, i propose click & active in issue.

If you think l,r,m is better anyway, i have no problem with da too.

**Comment by mouse0w0** [comment]
Yes, this makes block logic depend on key binding. I think I should give up this thinking.

**Comment by cvrunmin** [comment]
We should differ clearly what is 'click a block' and what is 'activate a block'

**Comment by mouse0w0** [comment]
As a general rule, 'click a block' is left click, 'activate a block' is right click, 'select a block' is middle click.

**Comment by cvrunmin** [comment]
Do we need `world` to be included in these events? A similar version of BlockClickEvent and BlockPlaceEvent is at `nullengine.event.world.block`

**Comment by mouse0w0** [comment]
Of course, this code is terrible.

File: EngineAPI/src/main/java/nullengine/block/event/BlockClickEvent.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import nullengine.entity.Entity;
+import nullengine.block.Block;
+import nullengine.event.Event;
+import nullengine.math.BlockPos;
+
+public class BlockClickEvent implements Event {
+    private BlockPos blockPos;
+    private Block block;
+    private Entity entity;
+
```
Comment (ci010) [inline EngineAPI/src/main/java/nullengine/block/event/BlockClickEvent.java]: Please do comment like this INLINE!

Like: there should be world reference in this event.

Comment (mouse0w0) [reply EngineAPI/src/main/java/nullengine/block/event/BlockClickEvent.java]: Your all commits have many troubles. I don't know what need comment inline!

Comment (ci010) [reply EngineAPI/src/main/java/nullengine/block/event/BlockClickEvent.java]: Then what trouble? You should write it down here other than just throw some "random judge". Nobody will understand that.

Comment (mouse0w0) [reply EngineAPI/src/main/java/nullengine/block/event/BlockClickEvent.java]: BlockClickEvent need include `World` field.

File: EngineAPI/src/main/java/nullengine/event/world/block/BlockInteractEvent.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+     * The hit vector should be in block local space
+     */
+    private Vector3fc hit;
+
+    private Facing face;
+
+    /**
+     * The cause contains the reason & the type of the interact.
+     */
+    private BlockInteractCause cause;
```
Comment (ci010) [inline EngineAPI/src/main/java/nullengine/event/world/block/BlockInteractEvent.java:32]: In this design, dev will determine whether this is a hit (left click) or a use (right click) by this field.

See entity trigger

File: EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockChangeCause.java [hunk 3]
```diff
@@ -1,7 +1,20 @@
...
 
+import nullengine.entity.Entity;
 import nullengine.event.cause.Cause;
+import nullengine.item.Item;
+import nullengine.world.World;
 
 public interface BlockChangeCause extends Cause {
+    interface EntityTrigger extends BlockChangeCause {
+        Entity getEntity();
+    }
```
Comment (ci010) [inline EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockChangeCause.java:11]: These are all design examples of how these things will be.

After this PR, the design of this should be clear.

Please have some suggestions @Mouse0w0  @cvrunmin

File: EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockInteractCause.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.util.Optional;
+
+import nullengine.entity.Entity;
+import nullengine.event.cause.Cause;
+import nullengine.item.ItemStack;
+
+public interface BlockInteractCause extends Cause {
+    class EntityTrigger implements BlockInteractCause {
+        public static final String HIT = "entity.hit", USE = "entity.use";
```
Comment (ci010) [inline EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockInteractCause.java:11]: Again, another way to do the cause.

Some suggestions please. @Mouse0w0  @cvrunmin 

Notice that inner class is just for convenience. In final design, we should consider where these things go

File: EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockChangeCause.java (hunk 3 above)
Comment (mouse0w0) [reply EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockChangeCause.java:11]: That's enough. I have nothing to say.

File: EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockInteractCause.java (hunk 4 above)
Comment (mouse0w0) [reply EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockInteractCause.java:11]: `HIT` and `USE` should move into the enum named `Action` of  `BlockInteractEvent`. And add a field named action to `BlockInteractEvent`.

Comment (cvrunmin) [reply EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockInteractCause.java:11]: I do prefer using class than using interface. Using interface may require creating classes implementing combinations of interfaces.

Comment (ci010) [reply EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockInteractCause.java:11]: > `HIT` and `USE` should move into the enum named `Action` of `BlockInteractEvent`. And add a field named action to `BlockInteractEvent`.

Just for extensibility requirement, if there are more than hit & use action, the enum is not a good way to do this. In this case, we just need a pivot to mark a semantic meaning of "Hit" & "Use". 

Either specific String, or immutable object (enum is a kind of immutable object without third-party extensibility) can play this role. 

I'm totally open on this.

If both of you think enum is simple and better. We just do the enum way.

Comment (ci010) [reply EngineAPI/src/main/java/nullengine/event/world/block/cause/BlockInteractCause.java:11]: > I do prefer using class than using interface. Using interface may require creating classes implementing combinations of interfaces.

Yes, BlockInteractCause.EntityTrigger is a class & you can create instance by calling function.

All other later work, like command triggered, will be wrapped in class. Interface is just the parent.

Refer [this](https://github.com/UnknownDomainGame/NullEngine/pull/44/files#diff-d7f35bed7bf563ff0f74311b10ec7610R151)

**Pull request closed without merging**